File: Python3/Add Binary.py
# 48ms, 100.0%
class Solution:
    def addBinary(self, a, b):
        if len(a) == 0: return b
        if len(b) == 0: return a
        if a[-1] == '1' and b[-1] == '1':
            return self.addBinary(self.addBinary(a[0: -1], b[0: -1]), '1') + '0'
        else:
            return self.addBinary(a[0: -1], b[0: -1]) + str(int(a[-1]) + int(b[-1]))
    

File: Java/Add Binary.java
// 1ms, 100.00%; 37.9MB, 5.62%
class Solution {
    public static String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length() - 1, j = b.length() - 1, carry = 0;
        while (i > -1 || j > -1) {
            if (i > -1) {
                carry += a.charAt(i--) - 48;
            }
            if (j > -1) {
                carry += b.charAt(j--) - 48;
            }
            sb.append(carry & 1);
            carry >>= 1;
        }
        if (carry == 1) {
            sb.append(carry);
        }
        return sb.reverse().toString();
    }
}
